dpcpp/base/executor.dp.cpp

```diff
@@ -0,0 +1,182 @@
+{
+    std::map<std::string, sycl::info::device_type> device_type_map{
+        {"accelerator", sycl::info::device_type::accelerator},
+        {"all", sycl::info::device_type::all},
+        {"cpu", sycl::info::device_type::cpu},
+        {"gpu", sycl::info::device_type::gpu}};
+    return sycl::device::get_devices(device_type_map.at(device_type));
+}
+
+
+}  // namespace detail
+
+
+void OmpExecutor::raw_copy_to(const DpcppExecutor *dest, size_type num_bytes,
+                              const void *src_ptr, void *dest_ptr) const
+{
+    if (num_bytes > 0) {
+        dest->get_queue()->memcpy(dest_ptr, src_ptr, num_bytes).wait();
+    }
+}
+
+
+std::shared_ptr<DpcppExecutor> DpcppExecutor::create(
+    int device_id, std::shared_ptr<Executor> master, std::string device_type)
+{
+    return std::shared_ptr<DpcppExecutor>(
+        new DpcppExecutor(device_id, std::move(master), device_type));
+}
+
+
+void DpcppExecutor::raw_free(void *ptr) const noexcept
+{
+    sycl::free(ptr, queue_->get_context());
+}
+
+
+void *DpcppExecutor::raw_alloc(size_type num_bytes) const
+{
+    void *dev_ptr = sycl::malloc_device(num_bytes, *queue_.get());
+    GKO_ENSURE_ALLOCATED(dev_ptr, "DPC++", num_bytes);
+    return dev_ptr;
+}
+
+
+void DpcppExecutor::raw_copy_to(const OmpExecutor *, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+{
+    if (num_bytes > 0) {
+        queue_->memcpy(dest_ptr, src_ptr, num_bytes).wait();
+    }
+}
+
+
+void DpcppExecutor::raw_copy_to(const CudaExecutor *dest, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+{
+    // TODO: later when possible, if we have DPC++ with a CUDA backend
+    // support/compiler, we could maybe support native copies?
+    GKO_NOT_SUPPORTED(dest);
+}
+
+
+void DpcppExecutor::raw_copy_to(const HipExecutor *dest, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+{
+    GKO_NOT_SUPPORTED(dest);
+}
+
+
+void DpcppExecutor::raw_copy_to(const DpcppExecutor *dest, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+{
+    if (num_bytes > 0) {
+        // TODO: does this work? Or is it needed to go through host?
+        dest->get_queue()->memcpy(dest_ptr, src_ptr, num_bytes).wait();
+    }
+}
+
+
+void DpcppExecutor::synchronize() const { queue_->wait_and_throw(); }
+
+
+void DpcppExecutor::run(const Operation &op) const
+{
+    this->template log<log::Logger::operation_launched>(this, &op);
+    op.run(std::static_pointer_cast<const DpcppExecutor>(
+        this->shared_from_this()));
+    this->template log<log::Logger::operation_completed>(this, &op);
+}
+
+
+int DpcppExecutor::get_num_devices(std::string device_type)
+{
+    return detail::get_devices(device_type).size();
+}
+
+
+void delete_queue(sycl::queue *queue)
+{
+    queue->wait();
+    delete queue;
```

Comment:
I guess we need this since std::unique_ptr would require a complete type in the header?

Reply:
Yes.